pageserver/src/layered_repository/delta_layer.rs

```diff
@@ -641,26 +786,134 @@ impl DeltaLayerWriter {
+        assert!(self.training);
+        self.sample_key_lsn_willinit.push((key, lsn, will_init));
+        self.sample_sizes.push(bytes.len());
+        self.sample_data.extend_from_slice(bytes);
+        Ok(())
+    }
+
+    /// Train the compression dictionary, and flush out all the accumulated
+    /// key-value pairs to disk.
+    fn finish_training(&mut self) -> Result<()> {
+        assert!(self.training);
+        assert!(self.sample_sizes.len() == self.sample_key_lsn_willinit.len());
+
+        // Create the dictionary, if we had enough samples for it.
+        //
+        // If there weren't enough samples, we don't do any compression at
+        // all. Possibly we could still benefit from compression; for example
+        // if you have only one gigantic value in a single layer, it would
+        // still be good to compress that, without a dictionary. But we don't
+        // do that currently.
+        if self.sample_sizes.len() >= config::ZSTD_MIN_SAMPLES {
+            let dictionary = zstd::dict::from_continuous(
+                &self.sample_data,
+                &self.sample_sizes,
+                config::ZSTD_MAX_DICTIONARY_SIZE,
+            )?;
+
+            let off = self.blob_writer.write_blob(&dictionary)?;
+            self.dictionary_offset = off;
+
+            let compressor = zstd::bulk::Compressor::with_dictionary(
+                config::ZSTD_COMPRESSION_LEVEL,
+                &dictionary,
+            )?;
+            self.compressor = Some(compressor);
+        };
+        self.training = false;
+
+        // release the memory used by the sample buffers
+        let sample_key_lsn_willinit = std::mem::take(&mut self.sample_key_lsn_willinit);
+        let sample_sizes = std::mem::take(&mut self.sample_sizes);
+        let sample_data = std::mem::take(&mut self.sample_data);
+
+        // Compress and write out all the buffered key-value pairs
+        let mut buf_idx: usize = 0;
+        for ((key, lsn, will_init), len) in
+            itertools::izip!(sample_key_lsn_willinit.iter(), sample_sizes.iter())
+        {
+            let end = buf_idx + len;
+            self.put_value_flush(*key, *lsn, *will_init, &sample_data[buf_idx..end])?;
+            buf_idx = end;
+        }
+        assert!(buf_idx == sample_data.len());
+
+        Ok(())
+    }
+
+    /// Write a key-value pair to the file, compressing it if applicable.
+    pub fn put_value_flush(
+        &mut self,
+        key: Key,
+        lsn: Lsn,
+        will_init: bool,
+        bytes: &[u8],
+    ) -> Result<()> {
+        assert!(!self.training);
         assert!(self.lsn_range.start <= lsn);
 
-        let off = self.blob_writer.write_blob(&Value::ser(&val)?)?;
+        let mut blob_content = bytes;
+        let mut compressed = false;
+
+        // Try to compress the blob
+        let compressed_bytes;
+        if let Some(ref mut compressor) = self.compressor {
+            compressed_bytes = compressor.compress(blob_content)?;
```

Comment:
I do not completely understand your comment. What do you mean by "page layer"?
Do you propose to hide compression logic in BlobWriter? But we need to store compressed tag in layer metadata.
Certainly it is possible to ask BlobWriter if blob was compressed. But I do not see much sense in it.

Reply:
Actually compression for all layers kind is done in different way:
-  we do not have compression for inmem layer because compressing data in memory on modern hardware has no sense. Yes, there is ephemeral file pages of which can be compressed. But it is only possible for page images which are rarely written to delta layer.
- For delta layer we have to use dictionary compression because size of wal record  may be too small to store dictionary with it. So training phase is required.
- For image layer we compress individually each page.

So in principle it is possible to hide all this logic somewhere. But as I wrote : i do not see much sense in it.

Reply:
> That in-file data could be compressed.

Not sure. It is temporary data which is expected to be removed soon. 
It may be not reasonable to spent CPU to compress/decompress this data.

> That doesn't have to be much different from the current design.

Sorry, I do not understand you. I have explained current implementation.

> I think we shouldn't be unable to read older layer files.

You mean layers written before compression was added?
I do not think that we should struggle here for backward compatibility at this moment.